server/src/main/java/org/elasticsearch/cluster/ClusterState.java

```diff
@@ -150,6 +151,8 @@ default boolean isPrivate() {
 
     private final DiscoveryNodes nodes;
 
+    private final Map<String, TransportVersion> transportVersions;
```

Comment:
Could we do this as a sorted list instead? We will already need to iterate over all elements in order to calculate the new minimum version. We only need to know where to add/remove a node from. And since ClusterState is immutable, we will have to copy all existing objects plus/minus the change anyways. So if we keep a sorted List, we can insert/filter into a new list on changes.

Reply:
Wouldn't it leave us open to potential problems if the discoverynodes collection & transportversions multiset got out of sync? Also it makes the structures a bit more difficult to inspect.

Reply:
I don't feel strongly about this, and defer to @DaveCTurner turner for everything cluster state related. However, my thinking is:
* Since cluster state is immutable, the read structure is more important than updates. We don't ever need to lookup the TransportVersion by node here. All we do is scan to find the lowest version. Even if we do that scan up front when creating the cluster state, we don't actually need the associative lookup provided by a Map.
* The Map doesn't make creating a new cluster state any easier. We still have to iterate over every element in the collection in order to make a copy. When we do that, we can filter out the node being removed.
* At minimum I would not use ImmutableOpenMap, just use the immutable tools from the jdk (either Map.of/copyOf or Collections.unmodifiableMap).

Reply:
We have full control over new cluster states being created, so it's on us to keep these things in sync. That said, I don't see much of a problem with using a `Map` here (plus a field to represent the pre-computed min version) although I think the JDK's immutable map will be fine.

Reply:
Changed to JDK maps